EIPS/eip-draft-blake2b-f-precompile.md

```diff
@@ -0,0 +1,76 @@
+discussions-to: https://ethereum-magicians.org/t/blake2b-f-precompile/3157
+status: Draft
+type: Standards Track
+category: Core
+created 2016-10-04
+requires: 2046
+---
+
+<!--You can leave these HTML comments in your merged EIP and delete the visible duplicate text guides, they will not appear and may be helpful to refer to if you edit it again. This is the suggested template for new EIPs. Note that an EIP number will be assigned by an editor. When opening a pull request to submit your EIP, please use an abbreviated title in the filename, `eip-draft_title_abbrev.md`. The title should be 44 characters or less.-->
+
+## Simple Summary
+<!--"If you can't explain it simply, you don't understand it well enough." Provide a simplified and layman-accessible explanation of the EIP.-->
+
+This EIP will enable the Blake2b hash function to run cheaply on the EVM, allowing easier interoperability between Ethereum and Zcash as well as other Equihash-based PoW coins.
+
+## Abstract
+<!--A short (~200 word) description of the technical issue being addressed.-->
+
+This EIP introduces a new precompiled contract which implements the compression function `F` used in the BLAKE2b cryptographic hashing algorithm, for the purpose of allowing interoperability between the EVM and Zcash, as well as introducing more flexible cryptographic hash primitives to the EVM.
+
+## Motivation
+<!--The motivation is critical for EIPs that want to change the Ethereum protocol. It should clearly explain why the existing protocol specification is inadequate to address the problem that the EIP solves. EIP submissions without sufficient motivation may be rejected outright.-->
+
+Besides being a useful cryptographic hash function and SHA3 finalist, BLAKE2b allows for efficient verification of the Equihash PoW used in Zcash, making a BTC Relay - style SPV client possible on Ethereum. A single verification of an Equihash PoW verification requires 512 iterations of the hash function, making verification of Zcash block headers prohibitively expensive if a Solidity implementation of BLAKE2b is used.
+
+The BLAKE2b algorithm is highly optimized for 64-bit CPUs, and is faster than MD5 on modern processors.
+
+Interoperability with Zcash could enable contracts like trustless atomic swaps between the chains, which could provide a much needed aspect of privacy to the very public Ethereum blockchain.
+
+## Specification
+<!--The technical specification should describe the syntax and semantics of any new feature. The specification should be detailed enough to allow competing, interoperable implementations for any of the current Ethereum platforms (go-ethereum, parity, cpp-ethereum, ethereumj, ethereumjs, and [others](https://github.com/ethereum/wiki/wiki/Clients)).-->
+
+Adds a precompile at address `0x0d` which accepts [ABI encoded](https://github.com/ethereum/wiki/wiki/Ethereum-Contract-ABI) arguments corresponding to the function signature
```

Comment:
Per our discussion, we should update the part about ABI encoded parameters.

Reply:
@pdyraga, I've been playing with your f function in go and it's great and you've done a fantastic job here - massive props. I have two questions but please do not answer the stupid ones: 
1) In F.go, should f return h? 
2) Should the package be the same name as the folder it's in?

Reply:
Another silly question: RFC7693 has 12 rounds but we're doing 10. I trust you guys but why 10 instead of 12?

Reply:
I recommend we move implementation discussion to the [blake2-f library's first PR](https://github.com/keep-network/blake2-f/pull/1)- WDYT?

Reply:
@carterpy Can you please repost your questions in the PR referenced by @mhluongo? I can happily answer them there.

Reply:
thanks pdyraga, questions reposted [here](https://github.com/keep-network/blake2-f/pull/1#issuecomment-503598145)

Reply:
Here is an alternative proposition about how to call the `F` function.

Instead of ABI-encoded parameters (no precompile currently do that) we can use a standard `staticcall`. 

I'd propose the following structure:
```
[ bytes32[0] | bytes32[1] ][ bytes32[2] | bytes32[3] | bytes32[4] | bytes32[5] ]
[      64 bytes for h     ][                    128 bytes for m                ]

[             bytes32[6]             ][             bytes32[7]             ]
[ 24 bytes padding | 8 bytes for t_0 ][ 24 bytes padding | 8 bytes for t_1 ]

[             bytes32[8]          ][              bytes32[9]              ]
[ 31 bytes padding | 1 byte for f ][ 28 bytes padding | 4 bytes for rounds]
```
where:
- `h` - state vector
- `m` - message block vector
- `t_0, t_1` - offset counter
- `f` - final block indicator flag
- `rounds` - number of rounds

This way we pack data in equal 32-byte (256-bit) slots, we are compatible with the rest of precompile calls and we do not require any bitwise juggling from callers.

For convenience, one may expose a wrapping function, similar to `F`:
```
    function F(bytes32[2] memory h, bytes32[4] memory m, uint64[2] memory t, bool f, uint32 rounds) public view returns (bytes32[2] memory) {
      bytes32[2] memory output;

      bytes32 _t0 = bytes32(uint256(t[0]));
      bytes32 _t1 = bytes32(uint256(t[1]));

      bytes32 _f;
      if (f) {
        _f = hex"0000000000000000000000000000000000000000000000000000000000000001";
      }

      bytes32 _rounds = bytes32(uint256(rounds));

      bytes32[10] memory args = [ h[0], h[1], m[0], m[1], m[2], m[3], _t0, _t1, _f, _rounds ];

      assembly {
            if iszero(staticcall(not(0), 0x09, args, 0x140, output, 0x40)) {
                revert(0, 0)
            }
        }
        return output;
    }

    function callF() public view returns (bytes32[2] memory) {
      bytes32[2] memory h;
      h[0] = hex"6a09e627f3bcc909bb67ae8484caa73b3c6ef372fe94b82ba54ff53a5f1d36f1";
      h[1] = hex"510e527fade682d19b05688c2b3e6c1f1f83d9abfb41bd6b5be0cd19137e2179";

      bytes32[4] memory m;
      m[0] = hex"278400340e6b05c5752592c52c4f121292eafc51cc01a997b4aed13409298fad";
      m[1] = hex"0d99ccc8f76453d9b3661e5c4d325d7751147db17489046d1682d50eefa4a1da";
      m[2] = hex"0000000000000000000000000000000000000000000000000000000000000000";
      m[3] = hex"0000000000000000000000000000000000000000000000000000000000000000";

      uint64[2] memory t;
      t[0] = 18446744073709551552;
      t[1] = 18446744073709551615;

      bool f = true;

      uint32 rounds = 12;

      return F(h, m, t, f, rounds);
    }
```

Reply:
(this looks terrible at the first sight, but this is similar to how other precompiles are called with assembly 🤷‍♀ - they are all equally terrible).